Services

taracell opened the pull request:
I believe I completed all requested things.  I cannot seem to simulate a stream or document.get() to produce the actual snapshot document for display or read.  As soon as firestore is enabled this will be better. I've written what I can with the actual calls the devs can make to firestore.

taracell committed:
Enable Firestore use

taracell committed:
Enable Firestore use

maddiew95 committed:
first commit in profile page

taracell committed:
Inital Services Commit

taracell committed:
Revisions per request. Moved test_screen.dart and renamed it to FirebaseCrashlytics_test_screen.dart.  Updated pubspec.yaml to latest release today as well as podfile update to firebase BOM 7.0.3 for iOS. Removed requested methods and data from main and placed into splash_page.dart.  If app has issues during main crashlytics and analytics will not log those issues from users.

taracell committed:
Merge remote-tracking branch 'origin/develop' into Services

taracell committed:
Revisions per Discord chat request from @Zuko 1/12/2021 09:10 AM PST

DanielQuick committed:
Change all usages of model to controller

maddiew95 committed:
updated from [develop] branch

taracell committed:
Updated to latest versions of Firebase/Firestore dependencies, fixed _initalize() per request placed comments and debugPrint...These may be taken out later if needed

maddiew95 committed:
made log out alert and profile status

taracell committed:
Attempt at providing requested solution

taracell committed:
Attempt at providing requested solution

taracell committed:
Merge remote-tracking branch 'origin/Services' into Services

# Conflicts:
#	lib/service/idea_service.dart
#	lib/service/sprint_service.dart
#	lib/service/user_service.dart

taracell committed:
Merge branch 'develop' into Services

taracell committed:
Merge remote-tracking branch 'origin/develop' into Services

# Conflicts:
#	README.md
#	android/.gitignore
#	android/app/build.gradle
#	ios/.gitignore
#	ios/Podfile
#	ios/Podfile.lock
#	ios/Runner.xcodeproj/project.pbxproj
#	lib/view/page/splash_page.dart
#	pubspec.lock
#	pubspec.yaml

manuelvargastapia commented:
I have a question. Shouldn't be the model transformations (`fromJson`, `toJson`) inside their respective model classes?

DanielQuick commented:
> I have a question. Shouldn't be the model transformations (`fromJson`, `toJson`) inside their respective model classes?

The best way to answer that is to ask yourself why you want those functions. In this case (in my opinion at least) the answer is to specifically communicate with Firebase.

Because models should be completely isolated from the backend, including knowing what format their data is stored in, I believe it is best to place that logic in the Service class.

The other option is to create a generic to/from json in the model and alter Firebase's json in the Service to match it. This would be the better choice if we were dealing with multiple backends, to have less code duplication. With only 1 backend it would end up adding slightly more code, not less.

manuelvargastapia commented:
It makes sense. Thanks

taracell committed:
Most Firestore working.  authentication_service.dart and idea_service.dart fully operational.  Did not notice SprintPosts model before.  Will work that issue tomorrow.

DanielQuick commented on lib/model/idea.dart:
```diff
@@ -1,3 +1,5 @@
+import 'package:flutter/material.dart';
```
Model files have no need to import the material library

DanielQuick commented on lib/model/sprint.dart:
```diff
@@ -1,3 +1,5 @@
+import 'package:flutter/material.dart';
```
Model files have no need to import the material library

DanielQuick commented on lib/model/user.dart:
```diff
@@ -1,16 +1,24 @@
+import 'package:flutter/cupertino.dart';
```
Model files have no need to import the cupertino library

DanielQuick commented on lib/model/user.dart:
```diff
@@ -1,16 +1,24 @@
-      username: username ?? this.username,
+      email: email ?? this.email,
+      userName: userName ?? this.userName,
+      photoURL: photoURL ?? "_",
```
if photoURL is not provided, it should be this.photoURL. Not "_"

DanielQuick replied:
Unless you have a good reason for this.

taracell replied:
Default string for photoURL should be something that we can attach an image to...I've found that it was easier to place a simple if statement to attach the default image to...if (doc.data()['photoURL')== _) {place default image}

DanielQuick replied:
Can we not check if the photoURL is null before placing the default image? Either way the copyWith function doesn't seem like the correct place to enforce it, it will overwrite the user's photoURL every chance it gets.

taracell replied:
I thought I changed this with the last commit...Maybe not but the line should read
```dart
photoURL: photoURL ?? this.photoURL ?? "_",
```

would that be acceptable?

DanielQuick commented on lib/service/idea_service.dart:
```diff
@@ -0,0 +1,193 @@
+      "titleArray": _idea.title.toLowerCase().split(new RegExp('\\s+')).toList(),
+      "description": _idea.description,
+      "createdAt": _idea.createdAt ?? DateTime.now().microsecondsSinceEpoch,
+      "updatedAt": _idea.updatedAt ?? DateTime.now().microsecondsSinceEpoch,
```
Use millisecondsSinceEpoch, not microseconds

taracell replied:
created a function in services.dart to calculate seconds since Epoch...so I can calculate a month

DanielQuick commented on lib/service/idea_service.dart:
```diff
@@ -0,0 +1,193 @@
+  Future<void> addToDB() async {
+    DocumentReference docReference = ideaRef.doc();
+    await ideaRef.doc(docReference.id).set(toJson(docReference.id));
+    _idea = _idea.copyWith(id: docReference.id, createdAt: DateTime.now().microsecondsSinceEpoch);
```
Use millisecondsSinceEpoch, not microseconds

DanielQuick commented on lib/service/idea_service.dart:
```diff
@@ -0,0 +1,193 @@
+  /// updates _idea udpatedAt in the database and updates _idea with that update
+  Future<void> updateUpdatedAt() async {
+    await ideaRef.doc(_idea.id).update({'updatedAt': DateTime.now().microsecondsSinceEpoch});
+    _idea = _idea.copyWith(updatedAt: DateTime.now().microsecondsSinceEpoch);
```
Use millisecondsSinceEpoch, not microseconds

DanielQuick commented on lib/service/idea_service.dart:
```diff
@@ -0,0 +1,193 @@
+  /// To get the document ID to pass to this function:
+  ///  1. StreamBuilder -> store as AsyncSnapshot<QuerySnapshot> snapshot using
+  ///    getAllIdeasFromDBStream(), getIdeasFromDBForCurrentMonthStream(), or searchIdeasByTitle()
+  ///  2. ListView -> children: snapshot.data.docs.map((document) widgets for each document)
+  ///  3. access a specific idea "onTapped" with this method using "document.id".
+  getIdeaFromDBDocument(String documentId) async{
+    DocumentSnapshot doc;
+    await ideaRef.doc(documentId).get().then((doc) {
+      if (doc.exists) {
+        print('Document data: ${doc.data()}');
+        _idea = _idea.copyWith(id: doc.id,
+            title: doc.data()["title"],
+            description: doc.data()["description"],
+            createdAt: doc.data()["createdAt"],
+            updatedAt: doc.data()["updatedAt"],
+            votes: doc.data()['categories'] as List<String>);
+      } else {
+        print('Document does not exist on the database');
+      }
+    });
+    ideaToString();
+    return doc;
+  }
+
+  ///get _idea as a stream, returns AsyncSnapshot<QuerySnapshot>
+  getCurrentIdeaFromDBAsStream() async{
+    debugPrint('getCurrentIdeaFromDBStream() performing...');
+    return ideaRef.doc(_idea.id).get().asStream();
+  }
+
+  ///get all idea documents from database,returns stream of AsyncSnapshot<QuerySnapshot>
+  getAllIdeasFromDBStream() async {
+    debugPrint('getAllIdeasFromDBStream() performing...');
+    return ideaRef.snapshots();
+  }
+
+  /// returns all ideas that were created within the last month as stream of AsyncSnapshot<QuerySnapshot>
+  getIdeasFromDBForCurrentMonthStream() async {
+    debugPrint('getIdeasFromDBForCurrentMonthStream() performing...');
+    return ideaRef.where('createdAt', isGreaterThanOrEqualTo: DateTime.now().microsecondsSinceEpoch - 2629743 );
```
Use millisecondsSinceEpoch, not microseconds

DanielQuick replied:
I imagine this number at the end represents a month. But if we run the function on Jan 2nd, we only want it to return results from Jan 1st onward. If we run it on Jan 13th it should return results from Jan 1st onwards. etc etc

DanielQuick commented on lib/service/sprint_service.dart:
```diff
@@ -0,0 +1,139 @@
+      "title": _sprint.title,
+      "titleArray": _sprint.title.toLowerCase().split(new RegExp('\\s+')).toList(),
+      "description": _sprint.description,
+      "createdAt": DateTime.now().microsecondsSinceEpoch,
```
Use millisecondsSinceEpoch, not microseconds

DanielQuick commented on lib/service/sprint_service.dart:
```diff
@@ -0,0 +1,139 @@
+  getUpdatedAt() {
+    return DateTime
+        .now()
+        .microsecondsSinceEpoch;
```
Use millisecondsSinceEpoch, not microseconds

DanielQuick commented on lib/service/idea_service.dart:
```diff
@@ -0,0 +1,193 @@
+            'createdAt: ${_idea.createdAt}, '
+            'updatedAt: ${_idea.updatedAt}, '
+            'votes: ${_idea.votes}');
+  }
```
This should be added to the model file - it does not care about the backend

DanielQuick replied:
Additionally in the model file it should only return the string value - not instantly print it. You can print it here in the service if need be.

DanielQuick commented on lib/service/idea_service.dart:
```diff
@@ -0,0 +1,193 @@
+import 'package:flutter/cupertino.dart';
```
Remove all updateXXX functions and replace with a single update({Idea idea}) function.

taracell replied:
If you do that you will re-write over entire idea.  I've tried this several times with firestore within other applications as well as this one.

DanielQuick replied:
You can have the controllers pass in a list of flags then, if you need to update each attribute individually.

update({Idea idea, List<IdeaAttribute> attributes}) where IdeaAttribute is an enum you can check against.

DanielQuick commented on lib/service/idea_service.dart:
```diff
@@ -0,0 +1,193 @@
+import 'package:flutter/cupertino.dart';
```
Remove the notion of a 'current idea'. Any function should take the idea as a parameter.

DanielQuick commented on lib/service/idea_service.dart:
```diff
@@ -0,0 +1,193 @@
+  }
+
+  /// adds the _idea to the database
+  Future<void> addToDB() async {
```
rename to

`Future<Idea> create({Idea idea}) async {`

Change implementation accordingly. Should return the generated Idea (I imagine the only difference will be the id field has a value)

DanielQuick commented on lib/service/idea_service.dart:
```diff
@@ -0,0 +1,193 @@
+  }
+
+  /// removes _idea from the database
+  Future<void> removeFromDB() async {
```
Change to

`Future<void> delete({Idea idea}) async {`

Update implementation accordingly

DanielQuick commented on lib/service/idea_service.dart:
```diff
@@ -0,0 +1,193 @@
+  /// returns all ideas that were created within the last month as stream of AsyncSnapshot<QuerySnapshot>
+  getIdeasFromDBForCurrentMonthStream() async {
+    debugPrint('getIdeasFromDBForCurrentMonthStream() performing...');
+    return ideaRef.where('createdAt', isGreaterThanOrEqualTo: DateTime.now().microsecondsSinceEpoch - 2629743 );
```
You can keep these as private functions, but what I want exposed to the controller developers are the following:

```
Future<Idea> get({String id}) async { // Gets the single Idea with id: id

Future<List<Idea>> getAll() async { // Gets all ideas for this month
```

taracell replied:
I do not understand what you mean by controller

DanielQuick replied:
The project is organized into Models, Views, Controllers, and Services.

Controllers will interact with the services you are writing. I would like the function definitions I wrote above to be exposed to the controllers (made public).

DanielQuick commented on lib/view/page/splash_page.dart:
```diff
@@ -31,11 +37,23 @@ class _SplashPageState extends State<SplashPage> {
       widget.onComplete();
     });
+
+    AuthenticationService _authenticationService = new AuthenticationService();
+    IdeaService _ideaService = new IdeaService();
+    SprintService _sprintService = new SprintService();
+    UserService _userService = new UserService();
```
Services should be registered as singletons in locator.dart

taracell replied:
What does this mean?

DanielQuick replied:
https://github.com/DanielQuick/Flutter-Sprints-Idea-Tracker/blob/develop/lib/locator.dart

locator.registerSingleton(AuthenticationService());

DanielQuick commented on lib/view/page/splash_page.dart:
```diff
@@ -31,11 +37,23 @@ class _SplashPageState extends State<SplashPage> {
+    IdeaService _ideaService = new IdeaService();
+    SprintService _sprintService = new SprintService();
+    UserService _userService = new UserService();
+    await _authenticationService.signUp("foobar@test.com", "123qweASD#\$%", "123qweASD#\$%");
+    await _authenticationService.signOut();
+    await _authenticationService.signIn("foobar@test.com", "123qweASD#\$%");
+    await _ideaService.runIdeaServiceTest();
```
Testing code should not be committed to the splash page

taracell replied:
I understand...I will place it in a test file once I can get everything working.

DanielQuick reviewed (changes requested):
Left reviews, I didn't check each service but left reviews for the IdeaService which can be applied to the others. I'll review again after you have made changes.

taracell committed:
Most Firestore working.

taracell committed:
Most Firestore working.

taracell committed:
Merge branch 'develop' into Services

taracell committed:
Most Firestore working.

taracell committed:
Most Firestore working.

taracell committed:
Merge remote-tracking branch 'origin/Services' into Services

# Conflicts:
#	lib/view/page/splash_page.dart

taracell committed:
Firestore finally working

taracell committed:
Firestore finally working

taracell committed:
Merge remote-tracking branch 'origin/Services' into Services

# Conflicts:
#	lib/service/authentication_service.dart

taracell committed:
Firestore finally working

taracell committed:
Firestore finally working

maddiew95 committed:
log out pop back to /landing page'

taracell committed:
Firestore finally working

taracell committed:
Firestore finally working

taracell committed:
Merge remote-tracking branch 'origin/Services' into Services

DanielQuick commented on lib/service/idea_service.dart:
```diff
@@ -1,109 +1,129 @@
 import 'package:flutter/cupertino.dart';
```
createdAt and updatedAt should be milliseconds since epoch. There are multiple places in this file you are assigning your 'seconds since epoch' value.

DanielQuick commented on lib/service/idea_service.dart:
```diff
@@ -113,81 +133,78 @@ class IdeaService {
-    return ideaRef.where('createdAt', isGreaterThanOrEqualTo: DateTime.now().microsecondsSinceEpoch - 2629743 );
+    return ideaRef.where('createdAt',
+        isGreaterThanOrEqualTo:
+            currentTimeInSeconds() - 2629743).snapshots();
```
This has the same issue as previously.

What this code will do on Jan 2nd: return results from Dec 2nd - Jan 2nd

What it should do on Jan 2nd: return results from Jan 1st - Jan 2nd.

Get the DateTime value of the first day of the month. Convert that to millisecondsSinceEpoch, compare to createdAt

DanielQuick reviewed (changes requested):
I believe I have responded to everything. Message me on Discord if you need more direction, I'll be more available this week.

taracell committed:
service function updates per requested changes almost all passing testing

taracell committed:
idea service function updates per requested changes

taracell committed:
Merge remote-tracking branch 'origin/Services' into Services

# Conflicts:
#	lib/service/idea_service.dart

maddiew95 committed:
changed requests 1 and 2

taracell committed:
all function updates per requested changes all passing testing
main was changed due to relocation of services calling in locator.  I attempted to create a file to hold this code but the application will not build without it

taracell committed:
all function updates per requested changes all passing testing
main was changed due to relocation of services calling in locator.  I attempted to create a file to hold this code but the application will not build without it

taracell committed:
Merge remote-tracking branch 'origin/Services' into Services

# Conflicts:
#	lib/service/sprint_service.dart

taracell committed:
Merge remote-tracking branch 'origin/Services' into Services

# Conflicts:
#	lib/service/sprint_service.dart

taracell committed:
Merge remote-tracking branch 'origin/Services' into Services

# Conflicts:
#	lib/service/sprint_service.dart

taracell committed:
Merge remote-tracking branch 'origin/profile-page' into Services

# Conflicts:
#	pubspec.lock

taracell committed:
Created reset password through e-mail and change password from within app

taracell committed:
took firestore init out of main

taracell committed:
revised per instructions...Corrected code in idea page to push idea, update  in Idea update controller needs revision...put example code there

taracell committed:
Merge branch 'develop' into Services

# Conflicts:
#	lib/locator.dart

taracell committed:
revised per instructions...Created methods to vote yes/no for Idea Object
Logged in user can only vote once per idea.
It programmatically increments, update voteYes/voteNo properties of the Idea object and it increments by 1 (use case for yes vote: _ideaService.update(idea, Update.vote,'yes').
Added creatorId to the idea object to make it easier to identify the creator of the idea, and inserted logic to only allow the creator to edit the description and title of the Idea object.

DanielQuick reviewed (changes requested):
I've only looked at the AuthenticationService so far, but here are a few changes required.

We need a way to get the authenticated user (the User object)
We need a way to know if we are signed in already

taracell reviewed (commented):
> I've only looked at the AuthenticationService so far, but here are a few changes required.
> 
> We need a way to get the authenticated user (the User object)
> We need a way to know if we are signed in already

We need a way to know if we are signed in already: 
I always force``` _authenticaitonServices.signOut();``` upon the open of the login_screen, login_page whatever it is called.  This ensure the current user of the application is the one signed in, so that there aren't any issues in firestore, and forces a re-establishment of authentication for the user.


To get the authenticated user (the User object) the flow of the logic is this: 
```dart
_authenticationService.signUp(String email, String password, String passwordVerify);
```
creates a User object from the authentication credentials and then calls: 
```dart
_userService.getUserDocument(user);
```
which returns a user document for use, and sets the current authenticated user to the static User user within the User class for use across the application.
To retrieve the current logged in user call:
```dart
  ///gets user from User Service
  Future<User> _getUser()async{
    UserService _userService = new UserService();
    User user = await _userService.getUser();
    print(user.id);
    return user;
  }
```
I use it like this within the Idea update function:
```dart
Future<Idea> update(Idea idea, UpdateIdea update, String updateString) async {
    User user = await _getUser();
    Idea updatedIdea = idea;
    switch (update) {
      case UpdateIdea.title:
        {
          if(user.id == idea.creatorId) {
            updatedIdea = await _updateTitle(idea, updateString);
          }
          return updatedIdea;
        }
        break;
... rest of switch
```

DanielQuick commented:
We need to know if we are logged in before we load the landing page. So we can bypass the landing page.

No service should have static variables. These are hiding any mistakes from not using the locator.

The User Service should not store any user variable. It should only get & update a user's public info.

The Authentication Service should have a stored user variable. This is the signed in user (null if not signed in). It should get the user's public info from the User Service and private info from Firebase's user.

taracell committed:
revised  per  instructions authentication_service.dart now has methods:
 authenticatedUser(): returns the current authenticated user from firestore document User object
 isSignedIn(): bool to let you know if the current user is signed in or not

taracell committed:
revised  per  instructions getAll() returns Future<List<Sprint>> in sprint_service.dart
getAll() returns Future<List<Idea>> in idea_service.dart.
setupFirebase() is now on splash_page.dart _initialize()
test_services.dart moved to test and testing logic commented out

taracell committed:
revised  per  instructions getAll() returns Future<List<Sprint>> in sprint_service.dart
getAll() returns Future<List<Idea>> in idea_service.dart.
setupFirebase() is now on splash_page.dart _initialize()
test_services.dart moved to test and testing logic commented out

taracell committed:
Merge remote-tracking branch 'origin/Services' into Services

# Conflicts:
#	lib/service/user_service.dart
#	test/test_services.dart

taracell committed:
Revised per request to add "userName" property to signUp() method
This will be the logged in user's display name, it cannot be the convention to sign up because Firestore will only take e-mail.
One way to use userName as signIn() is to persist the userName/email information on sharedPreferences on the local device then translate the userName to the email address for userName use when missing @ symbol during signIn() after signUp.

DanielQuick committed:
Change update function to take lists of attribute updates

DanielQuick committed:
Update main.dart

DanielQuick committed:
Merge branch 'develop' into Services

DanielQuick committed:
Remove profile page

DanielQuick committed:
Update Idea Details Page to use creatorId
